File: java/1032_StreamOfCharacters.java
class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean word = false;
    
    public TrieNode() {};
}


class StreamChecker {
    Deque<Character> letters;
    TrieNode trie;
    int maxLength;
    
    public StreamChecker(String[] words) {
        letters = new LinkedList<Character>();
        trie = new TrieNode();
        maxLength = 0;
        
        for (String w : words) {
            TrieNode node = trie;
            int wordLength = w.length();
            for (int i = wordLength-1; i >= 0; --i) {
                char c = w.charAt(i);
                if (!node.children.containsKey(c))
                    node.children.put(c, new TrieNode());
                node = node.children.get(c);
            }
            node.word = true;
            maxLength = w.length() > maxLength ? w.length() : maxLength;
        }
    }

    public boolean query(char l) {
        letters.addFirst(l);
        if (letters.size() > maxLength)
            letters.pollLast();
        
        TrieNode node = trie;
        for (char c : letters) {
            if (node.children.containsKey(c)) {
                node = node.children.get(c);
            } else {
                break;
            }
            
            if (node.word)
                return true;
        }
        return node.word;
            
    }
}

/**
 * Your StreamChecker object will be instantiated and called as such:
 * StreamChecker obj = new StreamChecker(words);
 * boolean param_1 = obj.query(letter);
 */

File: python/1032_StreamOfCharacters.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = False

class StreamChecker:
    def __init__(self, words: List[str]):
        self.trie = TrieNode()
        self.letters = deque()
        self.max_length = 0
        for w in words:
            node = self.trie
            for i in range(len(w)-1, -1, -1):
                c = w[i]
                if c not in node.children:
                    node.children[c] = TrieNode()
                node = node.children[c]
                
            node.word = True
            self.max_length = max(self.max_length, len(w))

    def query(self, l: str) -> bool:
        self.letters.appendleft(l)
        if len(self.letters) > self.max_length:
            self.letters.pop()
        
        node = self.trie
        for l in self.letters:
            if l in node.children:
                node = node.children[l]
            else:
                break
            if node.word:
                return True
        return node.word
            
# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)
